### PR TITLE
blog: post urbit precepts and precepts discussion

### PR DESCRIPTION
**Pull request by urcades**
Posting the split version of Precepts and its following Discussion

### CODE REVIEW

**Comment by urcades** [comment]
Hoping some of these AZ deep cut imageries are appreciated – Sedona is a wonderful town

File: content/blog/precepts-discussion.md [hunk 1]
```diff
@@ -0,0 +1,769 @@
++++
+title = "Precepts: Discussion"
+date = 2020-03-14
+description = "Urbit is a principled approach to system design and programming, but it's often not obvious what those technical principles are.  This is a list of the technical maxims that define Urbit's approach to engineering."
+[extra]
+author = "Philip Monk"
+ship = "~wicdev-wisret"
```
Comment (matildepark) [inline content/blog/precepts-discussion.md]: `~wicdev-wisryt`, pronounced "wicdev was right"

File: content/blog/precepts-discussion.md [hunk 2]
```diff
@@ -0,0 +1,769 @@
++++
+title = "Precepts: Discussion"
+date = 2020-03-14
+description = "Urbit is a principled approach to system design and programming, but it's often not obvious what those technical principles are.  This is a list of the technical maxims that define Urbit's approach to engineering."
```
Comment (matildepark) [inline content/blog/precepts-discussion.md]: try this, adjusted from the opener as a description for the discussion: `The precepts aren’t arguments. We discuss and justify them here.`

File: content/blog/precepts-discussion.md [hunk 3]
```diff
@@ -0,0 +1,769 @@
...
+
+A perfect static type system can guarantee `T` is a subset of `S` — if it typechecks, then it works. Obviously, if you let `T := {}` (ie no program typechecks), then this is vacuously true. So you want `T` to be as large as possible while still being a subset of `S` — as many programs as possible typecheck without them being incorrect. But it can't make `T == S` — there will always be some programs that do the correct thing in practice, but the type checker can't prove that.
+
+Dually, a perfect testing suite can guarantee that `S` is a subset of `T` — no correct program fails the tests. This is vacuously true if you have no tests, for then `T = U`. So you want `T` to be as small as possible while still being a superset of `S` — you don't want to reject valid programs if you can help it. But you can't make `T == S` — there will always be some programs that are invalid but pass your typechecks (at least, if your program requires a Turing machine).
+
+(It is possible to create a static type system that has `S` subset `T` property described above for a testing suite — see Erlang's Dialyzerand (related
+work)[http://user.it.uu.se/~tobiasl/publications/contracts.pdf].)
```
Comment (matildepark) [inline content/blog/precepts-discussion.md]: should be `[related work](http://user.it.uu.se/~tobiasl/publications/contracts.pdf)`

File: content/blog/precepts-discussion.md [hunk 4]
```diff
@@ -0,0 +1,769 @@
...
+
+> E5: Don't name anything you don't have to.
+
+> E6: If it's the same thing, give it the same name.
+
+It is often said that there are two hard things in computer science: cache invalidation and naming things. This is a good aphorism. As often, someone will respond with "There are two hard things in CS: cache invalidation, naming things, and off-by-one errors". This is a weak
+joke because off-by-one errors are not hard at all. They're mostly a matter of discipline, and there's a couple of conventions that help. If you're looping on a zero-indexed variable, use `i < num_loops`. If you're looping on a one-indexed variable, use `i <= num_loops`.
+
+However, cache invalidation and naming things are genuinely hard because they must be solved ad hoc. There is no one-size-fits-all cache invalidation strategy, and possibly there never will be. Similar for naming. So you must actually comprehend and work within each individual
+context.
```
Comment (matildepark) [inline content/blog/precepts-discussion.md:581]: Does the site deal well with the newline here? idk

File: content/blog/precepts.md [hunk 5]
```diff
@@ -0,0 +1,355 @@
++++
+title = "Precepts"
+date = 2020-03-14
+description = "Urbit is a principled approach to system design and programming, but it's often not obvious what those technical principles are.  This is a list of the technical maxims that define Urbit's approach to engineering."
```
Comment (matildepark) [inline content/blog/precepts.md]: Simplify for the Twitter cards, lol: `The technical maxims that define Urbit's approach to engineering.`

File: content/blog/precepts-discussion.md [hunk 6]
```diff
@@ -0,0 +1,769 @@
++++
+title = "Precepts: Discussion"
+date = 2020-03-14
```
Comment (matildepark) [inline content/blog/precepts-discussion.md]: These both should be adjusted for today's date.

File: content/blog/precepts.md [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+[extra]
+author = "Philip Monk"
+ship = "~wicdev-wisret"
```
Comment (matildepark) [inline content/blog/precepts.md]: See previous

**Review by matildepark — CHANGES_REQUESTED** [review]
Gonna look at this on the site now ...

File: content/blog/precepts-discussion.md (hunk 4 above)
Comment (matildepark) [reply content/blog/precepts-discussion.md:581]: It does, so resolved ~

**Comment by matildepark** [comment]
Blockquotes are unstyled; just wondering if you have an opinion on that ... I prefer it to excessive margins, but maybe we want to italicise them or provide some sort of style hint that it's a quote?

image

File: content/blog/precepts.md (hunk 7 above)
Comment (urcades) [reply content/blog/precepts.md]: fixed!

File: content/blog/precepts.md (hunk 5 above)
Comment (urcades) [reply content/blog/precepts.md]: good call

File: content/blog/precepts-discussion.md (hunk 1 above)
Comment (urcades) [reply content/blog/precepts-discussion.md]: lmfao, wonderful, ty

**Comment by urcades** [comment]
> Blockquotes are unstyled; just wondering if you have an opinion on that ... I prefer it to excessive margins, but maybe we want to italicise them or provide some sort of style hint that it's a quote?
> 
> image

This was how we had previous specified quotes to look:

Screen Shot 2020-03-18 at 1 00 22 PM

Is this something that'd be implemented via main.scss?

**Comment by matildepark** [comment]
Yeah that’s easy. border-left, padding-left, font-style: italic, on I think body.essay’s blockquotes — probably already in scss

—
~haddef-sigwen
https://urbit.org

**Comment by urcades** [comment]
> Yeah that’s easy. border-left, padding-left, font-style: italic, on I think body.essay’s blockquotes — probably already in scss — ~haddef-sigwen https://urbit.org
> […](#)
> On Mar 18, 2020, at 1:02 PM, édouard ***@***.***> wrote: Blockquotes are unstyled; just wondering if you have an opinion on that ... I prefer it to excessive margins, but maybe we want to italicise them or provide some sort of style hint that it's a quote? This was how we had previous specified quotes to look: Is this something that'd be implemented via main.scss? — You are receiving this because you were assigned. Reply to this email directly, view it on GitHub, or unsubscribe.

Is there an additional build step required for having changes in css propagate to the markdown'd pages? I've edited the right location (I believe), but I'm not seeing the styling change.

**Comment by matildepark** [comment]
Sometimes you gotta cancel and reserve the Zola build …

—
~haddef-sigwen
https://urbit.org

**Comment by matildepark** [comment]
Looks good.

**Comment by urcades** [comment]
Did a small fix to the bottom margin of post blockquotes, but just waiting on checks to complete. Will turn to Asana to ask for when to post. cc @crrncllns

**Comment by philipcmonk** [comment]
I do like the images!

**Comment by urcades** [comment]
There are some deep cuts in here from AZ, hope it is appreciated – Sedona
is a lovely place to source landscape imagery from.

—
~fabled-faster
https://urbit.org/

**Pull request merged**